pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```diff
@@ -269,9 +270,11 @@ public void run() {
 
                 try {
                     processResult(currentRecord, result);
-                } catch (Exception e) {
+                    result.join();
```

Comment:
This seems to have a huge impact on performance, no?

Reply:
Yes, but we need a way to interrogate the Future task to determine if an exception was thrown. Otherwise, we won't be able to tell if an exception occurred at all.

Reply:
@david-streamlio I would recommend changing JavaInstance.handeMessage():

https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java#L69

I don't think it is appropriate to always return a CompletableFuture<JavaExecutionResult>.  We do not need to return a CompletableFuture for functions that are not implemented to use async.  I would recommend  modifying to existing class JavaExecutionResult to have options on whether to return a completable future result or just the result.

Reply:
it is also pointless for create new CompletableFutures for non-async functions here:

https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java#L74

leads to more GC

Reply:
@jerrypeng Agreed on both points. However, we still need to address the async use case scenario. I am thinking that a shared data struct can be used by the Futures to communicate exceptions, and the main thread can check the data struct for the presence of any errors and if one is found raise the exception?